runhouse/rns/run_module_utils.py

```diff
@@ -14,80 +16,100 @@
-            runtime_env={"env_vars": {"PYTHONPATH": module_path or ""}},
+            runtime_env=runtime_env,
         )(logging_wrapped_fn)
-        if fn_type == "map":
-            obj_ref = [ray_fn.remote(arg, **kwargs) for arg in args]
-        elif fn_type == "starmap":
-            obj_ref = [ray_fn.remote(*arg, **kwargs) for arg in args]
-        elif fn_type == "queue" or fn_type == "remote":
-            obj_ref = ray_fn.remote(*args, **kwargs)
-        elif fn_type == "repeat":
-            [num_repeats, args] = args
-            obj_ref = [ray_fn.remote(*args, **kwargs) for _ in range(num_repeats)]
-        else:
-            raise ValueError(f"fn_type {fn_type} not recognized")
+        obj_ref = ray_fn.remote(fn_pointers, fn_type, *args, **kwargs)
 
         if fn_type == "remote":
             rh_config.obj_store.put_obj_ref(key=run_key, obj_ref=obj_ref)
-            res = run_key
+            res = pickle.dumps([run_key, None, None])
         else:
             res = ray.get(obj_ref)
     return res
 
 
-def get_fn_by_name(module_name, fn_name, relative_path=None):
-    if relative_path:
-        module_path = str((Path.home() / relative_path).resolve())
-        sys.path.append(module_path)
-        logger.info(f"Appending {module_path} to sys.path")
-
-    if module_name in rh_config.obj_store.imported_modules:
-        importlib.invalidate_caches()
-        rh_config.obj_store.imported_modules[module_name] = importlib.reload(
-            rh_config.obj_store.imported_modules[module_name]
-        )
-        logger.info(f"Reloaded module {module_name}")
+def get_fn_from_pointers(fn_pointers, fn_type, *args, **kwargs):
+    (module_path, module_name, fn_name) = fn_pointers
+    if module_name == "notebook":
+        fn = fn_name  # already unpickled
     else:
-        logger.info(f"Importing module {module_name}")
-        rh_config.obj_store.imported_modules[module_name] = importlib.import_module(
-            module_name
-        )
-    fn = getattr(rh_config.obj_store.imported_modules[module_name], fn_name)
-    return fn
+        if module_path:
+            sys.path.append(module_path)
+            logger.info(f"Appending {module_path} to sys.path")
+
+        if module_name in rh_config.obj_store.imported_modules:
+            importlib.invalidate_caches()
+            rh_config.obj_store.imported_modules[module_name] = importlib.reload(
+                rh_config.obj_store.imported_modules[module_name]
+            )
+            logger.info(f"Reloaded module {module_name}")
+        else:
+            logger.info(f"Importing module {module_name}")
+            rh_config.obj_store.imported_modules[module_name] = importlib.import_module(
+                module_name
+            )
+        fn = getattr(rh_config.obj_store.imported_modules[module_name], fn_name)
+
+    try:
+        if fn_type == "map":
+            result = [fn(arg, **kwargs) for arg in args]
```

Comment:
actually running this inside one large ray.remote will make it all run within the same process, so this mapping should be moved and handled outside of ray.remote

same with starmap